File: atlas-navigation-cpp/src/SLAMEngine.cpp
#include "atlas_navigation/SLAM/SLAMEngine.h"
#include <algorithm>
#include <cmath>
#include <chrono>

namespace atlas_navigation {
namespace slam {

SLAMEngine::SLAMEngine(const SLAMEngineConfig& config) : config_(config) {
    currentMap_.id = "map-" + std::to_string(nowMs());
    currentMap_.resolution = config_.resolution;
    currentMap_.timestamp = nowMs();
    currentEstimate_.timestamp = nowMs();
}

LocalMap SLAMEngine::processObservation(const Observation& observation) {
    observationCount_++;
    updateEstimate(observation);

    if (observation.type == "OBJECT_DETECTED") updateMapFromDetection(observation);
    else if (observation.type == "ODOMETRY") updateFromOdometry(observation);
    else if (observation.type == "POSE") updateFromPose(observation);

    if (shouldCreateKeyframe()) createKeyframe();

    if (config_.enableLoopClosure) {
        auto loopClosure = detectLoopClosure();
        if (loopClosure.detected) applyLoopClosure(loopClosure);
    }

    pruneObjects();
    currentMap_.timestamp = nowMs();
    return currentMap_;
}

void SLAMEngine::updateEstimate(const Observation& observation) {
    if (observation.type == "POSE" || observation.type == "ODOMETRY") {
        double alpha = computeUpdateWeight(observation.uncertainty);
        if (observation.data.position.has_value()) {
            auto pos = observation.data.position.value();
            currentEstimate_.position.x = currentEstimate_.position.x * (1 - alpha) + pos.x * alpha;
            currentEstimate_.position.y = currentEstimate_.position.y * (1 - alpha) + pos.y * alpha;
            currentEstimate_.position.z = currentEstimate_.position.z * (1 - alpha) + pos.z * alpha;
        }
        if (observation.data.orientation.has_value()) {
            currentEstimate_.orientation = observation.data.orientation.value();
        }
        if (observation.data.velocity.has_value()) {
            double dt = estimateDeltaTime();
            auto vel = observation.data.velocity.value();
            currentEstimate_.position.x += vel.x * dt;
            currentEstimate_.position.y += vel.y * dt;
            currentEstimate_.position.z += vel.z * dt;
            currentEstimate_.velocity = vel;
        }
        currentEstimate_.timestamp = observation.timestamp;
        currentEstimate_.confidence = std::max(0.0, std::min(1.0, 1.0 - observation.uncertainty));
    }
}

void SLAMEngine::updateMapFromDetection(const Observation& observation) {
    auto pos = observation.data.position.value_or(Vector3{0, 0, 0});
    auto it = std::find_if(currentMap_.objects.begin(), currentMap_.objects.end(),
        [&](const MapObject& obj) { return obj.label == observation.data.object; });

    if (it != currentMap_.objects.end()) {
        double k = observation.data.confidence;
        it->position.x = it->position.x * (1 - k) + pos.x * k;
        it->position.y = it->position.y * (1 - k) + pos.y * k;
        it->position.z = it->position.z * (1 - k) + pos.z * k;
        it->confidence = std::max(it->confidence, observation.data.confidence);
        it->lastSeen = nowMs();
    } else {
        MapObject obj;
        obj.id = "obj-" + std::to_string(currentMap_.objects.size() + 1);
        obj.label = observation.data.object;
        obj.position = pos;
        obj.confidence = observation.data.confidence;
        obj.lastSeen = nowMs();
        currentMap_.objects.push_back(obj);
    }
}

void SLAMEngine::updateFromOdometry(const Observation& observation) {
    double dt = estimateDeltaTime();
    if (observation.data.velocity.has_value()) {
        auto vel = observation.data.velocity.value();
        currentEstimate_.position.x += vel.x * dt;
        currentEstimate_.position.y += vel.y * dt;
        currentEstimate_.position.z += vel.z * dt;
        currentEstimate_.velocity = vel;
    }
    currentEstimate_.timestamp = observation.timestamp;
}

void SLAMEngine::updateFromPose(const Observation& observation) {
    if (observation.data.position.has_value()) {
        currentEstimate_.position = observation.data.position.value();
    }
    if (observation.data.orientation.has_value()) {
        currentEstimate_.orientation = observation.data.orientation.value();
    }
    if (observation.data.velocity.has_value()) {
        currentEstimate_.velocity = observation.data.velocity.value();
    }
    currentEstimate_.timestamp = observation.timestamp;
}

bool SLAMEngine::shouldCreateKeyframe() const {
    if (lastKeyframeId_.empty()) return true;
    auto it = keyframes_.find(lastKeyframeId_);
    if (it == keyframes_.end()) return true;

    double dx = currentEstimate_.position.x - it->second.estimate.position.x;
    double dy = currentEstimate_.position.y - it->second.estimate.position.y;
    double dz = currentEstimate_.position.z - it->second.estimate.position.z;
    return std::sqrt(dx*dx + dy*dy + dz*dz) > config_.keyframeDistanceThreshold;
}

void SLAMEngine::createKeyframe() {
    SLAMKeyframe kf;
    kf.id = "kf-" + std::to_string(observationCount_);
    kf.estimate = currentEstimate_;
    kf.objects = currentMap_.objects;
    kf.timestamp = nowMs();

    if (!lastKeyframeId_.empty()) {
        auto prevIt = keyframes_.find(lastKeyframeId_);
        if (prevIt != keyframes_.end()) {
            auto relTransform = computeRelativeTransform(prevIt->second.estimate, currentEstimate_);
            KeyframeConnection conn;
            conn.keyframeId = lastKeyframeId_;
            conn.relativePose = relTransform;
            conn.edgeType = "ODOMETRY";
            // Create default info matrix
            auto info = std::vector<std::vector<double>>(6, std::vector<double>(6, 0));
            for (int i = 0; i < 3; ++i) info[i][i] = 100.0;
            for (int i = 3; i < 6; ++i) info[i][i] = 10.0;
            conn.informationMatrix = info;
            kf.connections.push_back(conn);
            addPoseGraphEdge(lastKeyframeId_, kf.id, conn);
        }
    }

    keyframes_[kf.id] = kf;
    lastKeyframeId_ = kf.id;
}

LoopClosureResult SLAMEngine::detectLoopClosure() const {
    LoopClosureResult result;
    if (keyframes_.size() < 5) return result;

    auto currentIt = keyframes_.find(lastKeyframeId_);
    if (currentIt == keyframes_.end()) return result;

    std::vector<SLAMKeyframe> kfArray;
    for (const auto& [_, kf] : keyframes_) kfArray.push_back(kf);
    int searchWindow = std::min(10, static_cast<int>(kfArray.size()) - 1);

    for (int i = 0; i < searchWindow; ++i) {
        const auto& candidate = kfArray[kfArray.size() - 2 - i];
        if (candidate.id == lastKeyframeId_) continue;

        double score = computeLoopClosureScore(currentIt->second.objects, candidate.objects);
        if (score >= config_.loopClosureThreshold) {
            result.detected = true;
            result.keyframeId = candidate.id;
            result.transform = computeRelativeTransform(candidate.estimate, currentEstimate_);
            result.confidence = score;
            return result;
        }
    }
    return result;
}

double SLAMEngine::computeLoopClosureScore(
    const std::vector<MapObject>& currentObjects,
    const std::vector<MapObject>& candidateObjects) const {
    if (currentObjects.empty() || candidateObjects.empty()) return 0;

    int matches = 0;
    for (const auto& curObj : currentObjects) {
        for (const auto& candObj : candidateObjects) {
            if (curObj.label == candObj.label) {
                double dx = curObj.position.x - candObj.position.x;
                double dy = curObj.position.y - candObj.position.y;
                double dz = curObj.position.z - candObj.position.z;
                if (std::sqrt(dx*dx + dy*dy + dz*dz) < config_.resolution * 5) {
                    matches++;
                    break;
                }
            }
        }
    }
    return static_cast<double>(matches) / std::max(currentObjects.size(), candidateObjects.size());
}

void SLAMEngine::applyLoopClosure(const LoopClosureResult& loopClosure) {
    KeyframeConnection conn;
    conn.keyframeId = loopClosure.keyframeId;
    conn.relativePose = loopClosure.transform;
    conn.edgeType = "LOOP_CLOSURE";
    auto info = std::vector<std::vector<double>>(6, std::vector<double>(6, 0));
    for (int i = 0; i < 3; ++i) info[i][i] = 10.0;
    for (int i = 3; i < 6; ++i) info[i][i] = 1.0;
    conn.informationMatrix = info;

    auto currentIt = keyframes_.find(lastKeyframeId_);
    if (currentIt != keyframes_.end()) {
        currentIt->second.connections.push_back(conn);
    }

    addPoseGraphEdge(loopClosure.keyframeId, lastKeyframeId_, conn);
    optimizePoseGraph();
    loopClosureCount_++;
    lastLoopClosureTime_ = nowMs();
}

void SLAMEngine::addPoseGraphEdge(const std::string& fromId, const std::string& toId,
                                   const KeyframeConnection& connection) {
    poseGraph_[fromId][toId] = connection;
}

PoseGraphResult SLAMEngine::optimizePoseGraph() {
    if (keyframes_.size() < 3) {
        PoseGraphResult r;
        r.optimizedEstimate = currentEstimate_;
        r.converged = true;
        return r;
    }

    double startError = computePoseGraphError();
    int iterations = 0;
    bool converged = false;

    for (int i = 0; i < config_.optimizationIterations; ++i) {
        double error = computePoseGraphError();
        if (error < 0.001) { converged = true; break; }
        adjustKeyframes();
        iterations++;
    }

    double endError = computePoseGraphError();
    if (!lastKeyframeId_.empty()) {
        auto it = keyframes_.find(lastKeyframeId_);
        if (it != keyframes_.end()) currentEstimate_ = it->second.estimate;
    }

    PoseGraphResult r;
    r.optimizedEstimate = currentEstimate_;
    r.errorReduction = startError - endError;
    r.iterations = iterations;
    r.converged = converged;
    return r;
}

double SLAMEngine::computePoseGraphError() const {
    double totalError = 0;
    for (const auto& [fromId, connections] : poseGraph_) {
        auto fromIt = keyframes_.find(fromId);
        if (fromIt == keyframes_.end()) continue;
        for (const auto& [toId, conn] : connections) {
            auto toIt = keyframes_.find(toId);
            if (toIt == keyframes_.end()) continue;
            auto expected = computeRelativeTransform(fromIt->second.estimate, toIt->second.estimate);
            double dx = expected.position.x - conn.relativePose.position.x;
            double dy = expected.position.y - conn.relativePose.position.y;
            double dz = expected.position.z - conn.relativePose.position.z;
            totalError += dx*dx + dy*dy + dz*dz;
        }
    }
    return totalError;
}

void SLAMEngine::adjustKeyframes() {
    for (auto& [kfId, kf] : keyframes_) {
        auto connIt = poseGraph_.find(kfId);
        if (connIt == poseGraph_.end()) continue;

        double totalDx = 0, totalDy = 0, totalDz = 0, weightSum = 0;

        for (const auto& [toId, conn] : connIt->second) {
            auto toIt = keyframes_.find(toId);
            if (toIt == keyframes_.end()) continue;

            auto expected = computeRelativeTransform(kf.estimate, toIt->second.estimate);
            double dx = conn.relativePose.position.x - expected.position.x;
            double dy = conn.relativePose.position.y - expected.position.y;
            double dz = conn.relativePose.position.z - expected.position.z;

            // Use a default weight
            double weight = 1.0;
            totalDx += dx * weight;
            totalDy += dy * weight;
            totalDz += dz * weight;
            weightSum += weight;
        }

        if (weightSum > 0) {
            double scale = 0.1;
            kf.estimate.position.x += totalDx * scale / weightSum;
            kf.estimate.position.y += totalDy * scale / weightSum;
            kf.estimate.position.z += totalDz * scale / weightSum;
        }
    }
}

Pose SLAMEngine::computeRelativeTransform(const StateEstimate& from, const StateEstimate& to) const {
    Pose result;
    result.position.x = to.position.x - from.position.x;
    result.position.y = to.position.y - from.position.y;
    result.position.z = to.position.z - from.position.z;
    result.orientation = from.orientation;
    return result;
}

void SLAMEngine::pruneObjects() {
    int64_t now = nowMs();
    int64_t maxAge = config_.objectAgeThreshold;

    currentMap_.objects.erase(
        std::remove_if(currentMap_.objects.begin(), currentMap_.objects.end(),
            [&](const MapObject& obj) { return now - obj.lastSeen > maxAge; }),
        currentMap_.objects.end());

    if (static_cast<int>(currentMap_.objects.size()) > config_.maxObjects) {
        std::sort(currentMap_.objects.begin(), currentMap_.objects.end(),
            [](const MapObject& a, const MapObject& b) { return a.confidence < b.confidence; });
        currentMap_.objects.resize(static_cast<size_t>(config_.maxObjects));
    }
}

double SLAMEngine::computeUpdateWeight(double uncertainty) const {
    return std::max(0.01, std::min(0.99, 1.0 / (uncertainty + 1.0)));
}

double SLAMEngine::estimateDeltaTime() const {
    int64_t now = nowMs();
    return static_cast<double>(now - currentEstimate_.timestamp) / 1000.0;
}

LocalMap SLAMEngine::getMap() const { return currentMap_; }
StateEstimate SLAMEngine::getEstimate() const { return currentEstimate_; }

std::vector<SLAMKeyframe> SLAMEngine::getKeyframes() const {
    std::vector<SLAMKeyframe> result;
    for (const auto& [_, kf] : keyframes_) result.push_back(kf);
    return result;
}

SLAMEngine::PoseGraphStats SLAMEngine::getPoseGraphStats() const {
    PoseGraphStats stats;
    stats.vertexCount = static_cast<int>(keyframes_.size());
    int edgeCount = 0;
    for (const auto& [_, conns] : poseGraph_) edgeCount += static_cast<int>(conns.size());
    stats.edgeCount = edgeCount;
    stats.loopClosureCount = loopClosureCount_;
    return stats;
}

void SLAMEngine::clearMap() { currentMap_.objects.clear(); currentMap_.timestamp = nowMs(); }

void SLAMEngine::reset() {
    clearMap();
    keyframes_.clear();
    poseGraph_.clear();
    currentEstimate_ = StateEstimate{};
    currentEstimate_.timestamp = nowMs();
    lastKeyframeId_.clear();
    observationCount_ = 0;
    loopClosureCount_ = 0;
    lastLoopClosureTime_ = 0;
}

} // namespace slam
} // namespace atlas_navigation

File: atlas-runtime/Perception/SLAMEngine.ts
import { LocalMap, MapObject } from "../../atlas-kernel/Perception/LocalMap";
import { Observation, Vector3, StateEstimate } from "../../atlas-kernel/Perception/StateEstimate";

/**
 * SLAM Engine configuration
 */
export interface SLAMEngineConfig {
  /** Map resolution in meters per cell */
  resolution: number;
  /** Maximum number of objects to track */
  maxObjects: number;
  /** Object age threshold for pruning (ms) */
  objectAgeThreshold: number;
  /** Enable loop closure detection */
  enableLoopClosure: boolean;
  /** Loop closure confidence threshold */
  loopClosureThreshold: number;
  /** Keyframe distance threshold */
  keyframeDistanceThreshold: number;
  /** Pose graph optimization iterations */
  optimizationIterations: number;
}

/**
 * Default SLAM engine configuration
 */
const DEFAULT_SLAM_CONFIG: SLAMEngineConfig = {
  resolution: 0.1,
  maxObjects: 100,
  objectAgeThreshold: 30000,
  enableLoopClosure: true,
  loopClosureThreshold: 0.8,
  keyframeDistanceThreshold: 0.5,
  optimizationIterations: 50,
};

/**
 * Keyframe for pose graph
 */
export interface SLAMKeyframe {
  id: string;
  estimate: StateEstimate;
  objects: MapObject[];
  timestamp: number;
  connections: KeyframeConnection[];
}

/**
 * Connection between keyframes
 */
export interface KeyframeConnection {
  keyframeId: string;
  relativeTransform: Transform3D;
  confidence: number;
  isLoopClosure: boolean;
}

/**
 * 3D transformation
 */
export interface Transform3D {
  position: Vector3;
  orientation: {
    x: number;
    y: number;
    z: number;
    w: number;
  };
}

/**
 * Loop closure result
 */
export interface LoopClosureResult {
  detected: boolean;
  keyframeId: string;
  transform: Transform3D;
  confidence: number;
}

/**
 * Pose graph optimization result
 */
export interface PoseGraphResult {
  optimizedEstimate: StateEstimate;
  errorReduction: number;
  iterations: number;
  converged: boolean;
}

export class SLAMEngine {
  private currentMap: LocalMap;
  private keyframes: Map<string, SLAMKeyframe>;
  private poseGraph: Map<string, Map<string, KeyframeConnection>>;
  private currentEstimate: StateEstimate;
  private config: SLAMEngineConfig;
  private lastKeyframeId: string | null;
  private observationCount: number;
  private loopClosureCount: number;
  private lastLoopClosureTime: number;

  constructor(config?: Partial<SLAMEngineConfig>) {
    this.config = { ...DEFAULT_SLAM_CONFIG, ...config };

    this.currentMap = {
      id: `map-${Date.now()}`,
      objects: [],
      resolution: this.config.resolution,
      timestamp: Date.now(),
    };

    this.keyframes = new Map();
    this.poseGraph = new Map();
    this.currentEstimate = {
      position: { x: 0, y: 0, z: 0 },
      velocity: { x: 0, y: 0, z: 0 },
      orientation: { x: 0, y: 0, z: 0, w: 1 },
      confidence: 0,
      timestamp: Date.now(),
    };

    this.lastKeyframeId = null;
    this.observationCount = 0;
    this.loopClosureCount = 0;
    this.lastLoopClosureTime = 0;
  }

  /**
   * Processes an observation to update the local map.
   * This implements Simultaneous Localization and Mapping logic.
   */
  processObservation(observation: Observation): LocalMap {
    this.observationCount++;

    // Update current estimate based on observation type
    this.updateEstimate(observation);

    // Process based on observation type
    if (observation.type === "OBJECT_DETECTED") {
      this.updateMapFromDetection(observation);
    } else if (observation.type === "ODOMETRY") {
      this.updateFromOdometry(observation);
    } else if (observation.type === "POSE") {
      this.updateFromPose(observation);
    }

    // Check for keyframe creation
    if (this.shouldCreateKeyframe()) {
      this.createKeyframe();
    }

    // Check for loop closure
    if (this.config.enableLoopClosure) {
      const loopClosure = this.detectLoopClosure();
      if (loopClosure?.detected) {
        this.applyLoopClosure(loopClosure);
      }
    }

    // Prune old objects
    this.pruneObjects();

    this.currentMap.timestamp = Date.now();
    return { ...this.currentMap };
  }

  /**
   * Update state estimate from observation
   */
  private updateEstimate(observation: Observation): void {
    if (observation.type === "POSE" || observation.type === "ODOMETRY") {
      const data = observation.data;

      if ("position" in data) {
        // Direct pose update
        const newPosition = data.position as Vector3;
        const uncertainty = observation.uncertainty;

        // Weighted update
        const alpha = this.computeUpdateWeight(uncertainty);
        this.currentEstimate.position.x =
          this.currentEstimate.position.x * (1 - alpha) + newPosition.x * alpha;
        this.currentEstimate.position.y =
          this.currentEstimate.position.y * (1 - alpha) + newPosition.y * alpha;
        this.currentEstimate.position.z =
          this.currentEstimate.position.z * (1 - alpha) + newPosition.z * alpha;

        if ("orientation" in data) {
          this.currentEstimate.orientation = data.orientation as any;
        }

        this.currentEstimate.timestamp = observation.timestamp;
      } else if ("velocity" in data) {
        // Velocity update - integrate
        const dt = this.estimateDeltaTime();
        this.currentEstimate.position.x += (data as any).velocity.x * dt;
        this.currentEstimate.position.y += (data as any).velocity.y * dt;
        this.currentEstimate.position.z += (data as any).velocity.z * dt;
        this.currentEstimate.velocity = data as any;
      }

      // Update confidence based on observation uncertainty
      this.currentEstimate.confidence = Math.max(
        0,
        Math.min(1, 1 - observation.uncertainty)
      );
    }
  }

  /**
   * Update map from object detection
   */
  private updateMapFromDetection(observation: Observation): void {
    if (observation.type !== "OBJECT_DETECTED") return;

    const detectedObj = observation.data as { object: string; confidence: number; position?: Vector3 };
    const pos = detectedObj.position || { x: 0, y: 0, z: 0 };

    const existingObj = this.currentMap.objects.find(
      obj => obj.label === detectedObj.object
    );

    if (existingObj) {
      // Update existing object position using weighted average
      const k = detectedObj.confidence;
      existingObj.position.x = existingObj.position.x * (1 - k) + pos.x * k;
      existingObj.position.y = existingObj.position.y * (1 - k) + pos.y * k;
      existingObj.position.z = existingObj.position.z * (1 - k) + pos.z * k;
      existingObj.confidence = Math.max(existingObj.confidence, detectedObj.confidence);
      existingObj.lastSeen = Date.now();
    } else {
      // Add new object to map
      this.currentMap.objects.push({
        id: `obj-${this.currentMap.objects.length + 1}`,
        label: detectedObj.object,
        position: pos,
        confidence: detectedObj.confidence,
        lastSeen: Date.now(),
      });
    }
  }

  /**
   * Update from odometry observation
   */
  private updateFromOdometry(observation: Observation): void {
    const odometry = observation.data as {
      linear: Vector3;
      angular: Vector3;
    };

    const dt = this.estimateDeltaTime();

    // Integrate odometry
    this.currentEstimate.position.x += odometry.linear.x * dt;
    this.currentEstimate.position.y += odometry.linear.y * dt;
    this.currentEstimate.position.z += odometry.linear.z * dt;

    this.currentEstimate.velocity = odometry.linear;
    this.currentEstimate.timestamp = observation.timestamp;
  }

  /**
   * Update from direct pose observation
   */
  private updateFromPose(observation: Observation): void {
    const pose = observation.data as StateEstimate;

    this.currentEstimate.position = pose.position;
    this.currentEstimate.orientation = pose.orientation;
    this.currentEstimate.timestamp = observation.timestamp;

    if (pose.velocity) {
      this.currentEstimate.velocity = pose.velocity;
    }
  }

  /**
   * Check if a new keyframe should be created
   */
  private shouldCreateKeyframe(): boolean {
    if (!this.lastKeyframeId) return true;

    const lastKeyframe = this.keyframes.get(this.lastKeyframeId);
    if (!lastKeyframe) return true;

    const dx = this.currentEstimate.position.x - lastKeyframe.estimate.position.x;
    const dy = this.currentEstimate.position.y - lastKeyframe.estimate.position.y;
    const dz = this.currentEstimate.position.z - lastKeyframe.estimate.position.z;

    const distance = Math.sqrt(dx * dx + dy * dy + dz * dz);

    return distance > this.config.keyframeDistanceThreshold;
  }

  /**
   * Create a new keyframe
   */
  private createKeyframe(): void {
    const keyframeId = `kf-${this.observationCount}`;

    const keyframe: SLAMKeyframe = {
      id: keyframeId,
      estimate: { ...this.currentEstimate },
      objects: [...this.currentMap.objects],
      timestamp: Date.now(),
      connections: [],
    };

    this.keyframes.set(keyframeId, keyframe);

    // Add connection to previous keyframe if exists
    if (this.lastKeyframeId) {
      const prevKeyframe = this.keyframes.get(this.lastKeyframeId);
      if (prevKeyframe) {
        const relativeTransform = this.computeRelativeTransform(
          prevKeyframe.estimate,
          this.currentEstimate
        );

        const connection: KeyframeConnection = {
          keyframeId: this.lastKeyframeId,
          relativeTransform,
          confidence: this.currentEstimate.confidence,
          isLoopClosure: false,
        };

        keyframe.connections.push(connection);

        // Add to pose graph
        this.addPoseGraphEdge(this.lastKeyframeId, keyframeId, connection);
      }
    }

    this.lastKeyframeId = keyframeId;
  }

  /**
   * Detect loop closure with previous keyframes
   */
  private detectLoopClosure(): LoopClosureResult | null {
    if (this.keyframes.size < 5) return null;

    const currentKeyframe = this.keyframes.get(this.lastKeyframeId!);
    if (!currentKeyframe) return null;

    // Check recent keyframes for potential loop closure
    const keyframeArray = Array.from(this.keyframes.values());
    const searchWindow = Math.min(10, keyframeArray.length - 1);

    for (let i = 0; i < searchWindow; i++) {
      const candidateKeyframe = keyframeArray[keyframeArray.length - 2 - i];

      // Skip if too recent
      if (candidateKeyframe.id === this.lastKeyframeId) continue;

      // Check for object matches
      const matchScore = this.computeLoopClosureScore(
        currentKeyframe.objects,
        candidateKeyframe.objects
      );

      if (matchScore >= this.config.loopClosureThreshold) {
        const transform = this.computeRelativeTransform(
          candidateKeyframe.estimate,
          this.currentEstimate
        );

        return {
          detected: true,
          keyframeId: candidateKeyframe.id,
          transform,
          confidence: matchScore,
        };
      }
    }

    return null;
  }

  /**
   * Compute loop closure score based on object matching
   */
  private computeLoopClosureScore(
    currentObjects: MapObject[],
    candidateObjects: MapObject[]
  ): number {
    if (currentObjects.length === 0 || candidateObjects.length === 0) {
      return 0;
    }

    let matches = 0;

    for (const currentObj of currentObjects) {
      for (const candidateObj of candidateObjects) {
        if (currentObj.label === candidateObj.label) {
          // Check if positions are consistent
          const dx = currentObj.position.x - candidateObj.position.x;
          const dy = currentObj.position.y - candidateObj.position.y;
          const dz = currentObj.position.z - candidateObj.position.z;

          const distance = Math.sqrt(dx * dx + dy * dy + dz * dz);

          if (distance < this.config.resolution * 5) {
            matches++;
            break;
          }
        }
      }
    }

    return matches / Math.max(currentObjects.length, candidateObjects.length);
  }

  /**
   * Apply loop closure constraint
   */
  private applyLoopClosure(loopClosure: LoopClosureResult): void {
    // Add loop closure edge to pose graph
    const connection: KeyframeConnection = {
      keyframeId: loopClosure.keyframeId,
      relativeTransform: loopClosure.transform,
      confidence: loopClosure.confidence,
      isLoopClosure: true,
    };

    const currentKeyframe = this.keyframes.get(this.lastKeyframeId!);
    if (currentKeyframe) {
      currentKeyframe.connections.push(connection);
    }

    this.addPoseGraphEdge(loopClosure.keyframeId, this.lastKeyframeId!, connection);

    // Optimize pose graph
    this.optimizePoseGraph();

    this.loopClosureCount++;
    this.lastLoopClosureTime = Date.now();
  }

  /**
   * Add edge to pose graph
   */
  private addPoseGraphEdge(
    fromId: string,
    toId: string,
    connection: KeyframeConnection
  ): void {
    if (!this.poseGraph.has(fromId)) {
      this.poseGraph.set(fromId, new Map());
    }
    this.poseGraph.get(fromId)!.set(toId, connection);
  }

  /**
   * Optimize the pose graph
   */
  private optimizePoseGraph(): PoseGraphResult {
    if (this.keyframes.size < 3) {
      return {
        optimizedEstimate: { ...this.currentEstimate },
        errorReduction: 0,
        iterations: 0,
        converged: true,
      };
    }

    const startError = this.computePoseGraphError();
    let iterations = 0;
    let converged = false;

    // Simple gradient descent optimization
    for (let i = 0; i < this.config.optimizationIterations; i++) {
      const error = this.computePoseGraphError();

      if (error < 0.001) {
        converged = true;
        break;
      }

      // Adjust keyframe estimates based on constraints
      this.adjustKeyframes();

      iterations++;
    }

    const endError = this.computePoseGraphError();

    // Update current estimate from last keyframe
    if (this.lastKeyframeId) {
      const lastKeyframe = this.keyframes.get(this.lastKeyframeId);
      if (lastKeyframe) {
        this.currentEstimate = { ...lastKeyframe.estimate };
      }
    }

    return {
      optimizedEstimate: { ...this.currentEstimate },
      errorReduction: startError - endError,
      iterations,
      converged,
    };
  }

  /**
   * Compute total error in pose graph
   */
  private computePoseGraphError(): number {
    let totalError = 0;

    for (const [fromId, connections] of this.poseGraph) {
      const fromKeyframe = this.keyframes.get(fromId);
      if (!fromKeyframe) continue;

      for (const [toId, connection] of connections) {
        const toKeyframe = this.keyframes.get(toId);
        if (!toKeyframe) continue;

        // Compute expected transform
        const expectedTransform = this.computeRelativeTransform(
          fromKeyframe.estimate,
          toKeyframe.estimate
        );

        // Compute error
        const dx = expectedTransform.position.x - connection.relativeTransform.position.x;
        const dy = expectedTransform.position.y - connection.relativeTransform.position.y;
        const dz = expectedTransform.position.z - connection.relativeTransform.position.z;

        totalError += dx * dx + dy * dy + dz * dz;
      }
    }

    return totalError;
  }

  /**
   * Adjust keyframe estimates to satisfy constraints
   */
  private adjustKeyframes(): void {
    const keyframeIds = Array.from(this.keyframes.keys());

    // Simple adjustment: move each keyframe towards constraint satisfaction
    for (const keyframeId of keyframeIds) {
      const connections = this.poseGraph.get(keyframeId);
      if (!connections) continue;

      const keyframe = this.keyframes.get(keyframeId);
      if (!keyframe) continue;

      let totalDx = 0;
      let totalDy = 0;
      let totalDz = 0;
      let weightSum = 0;

      for (const [toId, connection] of connections) {
        const toKeyframe = this.keyframes.get(toId);
        if (!toKeyframe) continue;

        const expectedTransform = this.computeRelativeTransform(
          keyframe.estimate,
          toKeyframe.estimate
        );

        const dx = connection.relativeTransform.position.x - expectedTransform.position.x;
        const dy = connection.relativeTransform.position.y - expectedTransform.position.y;
        const dz = connection.relativeTransform.position.z - expectedTransform.position.z;

        const weight = connection.confidence;
        totalDx += dx * weight;
        totalDy += dy * weight;
        totalDz += dz * weight;
        weightSum += weight;
      }

      if (weightSum > 0) {
        const scale = 0.1; // Learning rate
        keyframe.estimate.position.x += totalDx * scale;
        keyframe.estimate.position.y += totalDy * scale;
        keyframe.estimate.position.z += totalDz * scale;
      }
    }
  }

  /**
   * Compute relative transform between two poses
   */
  private computeRelativeTransform(
    from: StateEstimate,
    to: StateEstimate
  ): Transform3D {
    return {
      position: {
        x: to.position.x - from.position.x,
        y: to.position.y - from.position.y,
        z: to.position.z - from.position.z,
      },
      orientation: from.orientation, // Simplified
    };
  }

  /**
   * Prune old/dead objects from the map
   */
  private pruneObjects(): void {
    const now = Date.now();
    const maxAge = this.config.objectAgeThreshold;

    this.currentMap.objects = this.currentMap.objects.filter(
      obj => now - obj.lastSeen < maxAge
    );

    // Limit total objects
    if (this.currentMap.objects.length > this.config.maxObjects) {
      // Remove lowest confidence objects
      this.currentMap.objects.sort((a, b) => a.confidence - b.confidence);
      this.currentMap.objects = this.currentMap.objects.slice(0, this.config.maxObjects);
    }
  }

  /**
   * Compute update weight from uncertainty
   */
  private computeUpdateWeight(uncertainty: number): number {
    // Lower uncertainty = higher weight
    return Math.max(0.01, Math.min(0.99, 1 / (uncertainty + 1)));
  }

  /**
   * Estimate delta time from timestamps
   */
  private estimateDeltaTime(): number {
    const now = Date.now();
    const dt = now - this.currentEstimate.timestamp;
    return dt / 1000; // Convert to seconds
  }

  /**
   * Get current map
   */
  getMap(): LocalMap {
    return { ...this.currentMap };
  }

  /**
   * Get current state estimate
   */
  getEstimate(): StateEstimate {
    return { ...this.currentEstimate };
  }

  /**
   * Get all keyframes
   */
  getKeyframes(): SLAMKeyframe[] {
    return Array.from(this.keyframes.values());
  }

  /**
   * Get pose graph statistics
   */
  getPoseGraphStats(): {
    vertexCount: number;
    edgeCount: number;
    loopClosureCount: number;
  } {
    let edgeCount = 0;
    for (const connections of this.poseGraph.values()) {
      edgeCount += connections.size;
    }

    return {
      vertexCount: this.keyframes.size,
      edgeCount,
      loopClosureCount: this.loopClosureCount,
    };
  }

  /**
   * Clear the map and reset
   */
  clearMap(): void {
    this.currentMap.objects = [];
    this.currentMap.timestamp = Date.now();
  }

  /**
   * Reset the SLAM engine
   */
  reset(): void {
    this.clearMap();
    this.keyframes.clear();
    this.poseGraph.clear();
    this.currentEstimate = {
      position: { x: 0, y: 0, z: 0 },
      velocity: { x: 0, y: 0, z: 0 },
      orientation: { x: 0, y: 0, z: 0, w: 1 },
      confidence: 0,
      timestamp: Date.now(),
    };
    this.lastKeyframeId = null;
    this.observationCount = 0;
    this.loopClosureCount = 0;
    this.lastLoopClosureTime = 0;
  }
}
